src/charm.py

```diff
@@ -1,104 +1,456 @@
-                    "environment": {"thing": self.model.config["thing"]},
-                }
-            },
-        }
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ActiveStatus()
-
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
-
-        Learn more about config at https://juju.is/docs/sdk/config
+        if not self.unit.is_leader():
+            # NOTE: In cases where leader status is lost before the charm is
+            # finished processing all teardown events, this prevents teardown
+            # event code from running. Luckily, for this charm, none of the
+            # teardown code is necessary to perform if we're removing the
+            # charm.
+            event.add_status(BlockedStatus("Scaling is not implemented for this charm"))
+            logger.info("Scaling is not implemented for this charm")
+            return
+        try:
+            self._charm_config: CharmConfig = CharmConfig.from_charm(charm=self)  # type: ignore[no-redef]  # noqa: E501
+        except CharmConfigInvalidError as exc:
+            event.add_status(BlockedStatus(exc.msg))
+            return
+        if not self._relation_created(N2_RELATION_NAME):
+            event.add_status(BlockedStatus("Waiting for N2 relation to be created"))
+            logger.info("Waiting for N2 relation to be created")
+            return
+        if not self._container.can_connect():
+            event.add_status(WaitingStatus("Waiting for container to be ready"))
+            logger.info("Waiting for container to be ready")
+            return
+        if not self._statefulset_is_patched():
+            event.add_status(WaitingStatus("Waiting for statefulset to be patched"))
+            logger.info("Waiting for statefulset to be patched")
+            return
+        self.unit.set_workload_version(self._get_workload_version())
+        if not self._container.exists(path=BASE_CONFIG_PATH):
+            event.add_status(WaitingStatus("Waiting for storage to be attached"))
+            logger.info("Waiting for storage to be attached")
+            return
+        if not self._n2_requirer.amf_hostname:
+            event.add_status(WaitingStatus("Waiting for N2 information"))
+            logger.info("Waiting for N2 information")
+            return
+        event.add_status(ActiveStatus())
+
+    def _on_install(self, _) -> None:
```

Comment:
We do check that in the `_configure` method, but I treat it more as a backup. I don't see a reason to wait for all other checks to be satisfied before patching the statefulset if it can be done right off the bat.

Reply:
Humm, I'm not convinced, I'd go for simplicity here and remove the hook handler. If we find it takes too long to deploy we should optimize then.